Mathlib/Computability/ContextFreeGrammar.lean

```diff
@@ -0,0 +1,124 @@
+/-
+Copyright (c) 2023 Martin Dvorak. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Martin Dvorak
+-/
+import Mathlib.Computability.Language
+
+/-!
+# Context-Free Grammar
+
+This file contains the definition of a Context-Free Grammar, which is a grammar that has a single
+nonterminal symbol on the left-hand side of each rule.
+-/
+
+/-- Rule that rewrites a single nonterminal to any list of symbols. -/
+structure ContextFreeRule (T : Type*) (N : Type*) where
+  /-- Input nonterminal a.k.a. left-hand side -/
+  input : N
+  /-- Output string a.k.a. right-hand side -/
+  output : List (Symbol T N)
+
+/-- Context-free grammar that generates words over the alphabet `T` (a type of terminals). -/
+structure ContextFreeGrammar (T : Type*) where
+  /-- Type of nonterminals -/
+  NT : Type*
```

Comment:
Note there is a subtlety here about universes. It might be clearest to write
```suggestion
structure ContextFreeGrammar.{uNT uT} (T : Type uT) where
  /-- Type of nonterminals -/
  NT : Type uNT
```
which means almost exactly the same thing, but then "the type of CFGs with non-terminal symbols in universe `w`" is spelt `ContextFreeGrammar.{w} T` instead of `ContextFreeGrammar.{_ w} T`

Reply:
I don't want this.

Reply:
Which bit don't you want? All my change does is remove the need for the underscore

Reply:
I don't want to clutter my code so much. My suggestion is to revert everything in this file to `Type`. Nobody wants to work with grammars over higher universes anyways.

Reply:
Your code has always supported grammars over higher universes: the comparison is:

* `Type _` let lean choose the universes for you by magic
* `Type*` let lean choose the universe *names* for you by enumeration (`u_1, u_2`, ...)
* `Type uNT`: name them yourself

Any time you put a type within a structure (which is a reasonable choice here), you have more universe variable issues to worry about than usual.

Reply:
Note that this clutter can be reduced by removing the `{}` and putting it on a `universes` line.

Reply:
I pushed the change I was suggesting; it also makes the goal states easier to read since now you see `T : Type uT` instead of `T : Type u_1`.

Reply:
Thanks for the commit! I am happy that it now works.

Nevertheless, I'd prefer the old version that allowed `Type` only. The code is significantly more cluttered now.